dig.go

```diff
@@ -57,10 +57,47 @@ type optionFunc func(*Container)
 
 func (f optionFunc) applyOption(c *Container) { f(c) }
 
-// A ProvideOption modifies the default behavior of Provide. It's included for
-// future functionality; currently, there are no concrete implementations.
+type provideOptions struct {
+	Name string
+}
+
+// A ProvideOption modifies the default behavior of Provide.
 type ProvideOption interface {
-	unimplemented()
+	applyProvideOption(*provideOptions)
+}
+
+type provideOptionFunc func(*provideOptions)
+
+func (f provideOptionFunc) applyProvideOption(opts *provideOptions) { f(opts) }
+
+// Name is a ProvideOption that specifies that all values produced by a
+// constructor should have the given name. See also the package documentation
+// about Named Values.
+//
+// The name also applies to the fields of a dig.Out struct returned by the
+// constructor if those fields did not have their own name:"..." tag.
+//
+// Given,
+//
+//   type ConnectionResult struct {
+//     dig.Out
+//
+//     Conn    *sql.DB
+//     ROConn  *sql.DB `name:"ro"`
+//   }
+//
+//   func NewConnection(...) (ConnectionResult, error)
+//
+// The following will provide two connections to the container: one under the
+// name "ro" and the other under the name "rw".
+//
+//   c.Provide(NewConnection, dig.Name("rw"))
+//
+// This option has no effect on Value Groups.
+func Name(name string) ProvideOption {
```

Comment:
Not sure if we want to reuse the "Name" in other Options, would need to make an interface/struct to make that work.

Reply:
Discussed offline but: Even if we could support a similar option for
parameters, it would have to be a new option rather than `Name` being re-used
because they would both need to be provided at `c.Provide(..)` time. That
option could be a superset of ProvideOption and InvokeOption if needed.

I don't foresee us wanting to re-use this option but if we find a really
compelling need in the future, two things worth noting:

- Cutting a 2.0 of Dig is cheap as long as the Fx API remains the same.
- Depending on where you draw the line on what constitutes as breaking at this
  level, changing `Name` (a top-level function) to return a type that is valid
  for both, its current return type and something else *should* be okay.